homework/shared_ptr/weak_ptr.hpp

```diff
@@ -0,0 +1,128 @@
+#include <iostream>
+
+#include "sharedControlBlock.hpp"
+#include "shared_ptr.hpp"
+
+namespace cs {
+
+template <typename T>
+class weak_ptr {
+public:
+    constexpr weak_ptr() noexcept = default;
+    weak_ptr(const weak_ptr& r) noexcept;
+    weak_ptr(const cs::shared_ptr<T>& r) noexcept;
+    weak_ptr(weak_ptr&& r) noexcept;
+
+    weak_ptr& operator=(const weak_ptr& r) noexcept;
+    weak_ptr& operator=(const shared_ptr<T>& r) noexcept;
+    weak_ptr& operator=(weak_ptr&& r) noexcept;
+    ~weak_ptr();
+    size_t use_count() const noexcept;
+    bool expired() const noexcept;
+    cs::shared_ptr<T> lock() const noexcept;
+    void reset() noexcept;
+
+private:
+    T* ptr_ = nullptr;
+    SharedControlBlock<T>* controlBlock_ = nullptr;
+};
+
+template <typename T>
+weak_ptr<T>::weak_ptr(const weak_ptr<T>& r) noexcept
+    : ptr_{r.ptr_}, controlBlock_{r.controlBlock_} {
+    if (controlBlock_) {
+        controlBlock_->incrementWeakRefs();
+    }
+}
+
+template <typename T>
+weak_ptr<T>::weak_ptr(const cs::shared_ptr<T>& r) noexcept
+    : ptr_{r.ptr_}, controlBlock_{r.controlBlock_} {
+    controlBlock_->incrementWeakRefs();
+}
+
+template <typename T>
+weak_ptr<T>::weak_ptr(weak_ptr<T>&& r) noexcept
+    : ptr_{std::move(r.ptr_)}, controlBlock_{std::move(r.controlBlock_)} {
+    r.ptr_ = nullptr;
+    r.controlBlock_ = nullptr;
+}
+
+template <typename T>
+weak_ptr<T>& weak_ptr<T>::operator=(const weak_ptr<T>& r) noexcept {
+    ptr_ = r.ptr_;
+    if (controlBlock_) {
+        controlBlock_->incrementWeakRefs();
+    }
+    return *this;
+}
+
```

Comment:
You should add controlBlock from the original pointer to new one

Reply:
You should check that we don't make self-assignment